src/peergos/server/storage/S3Request.java

```diff
@@ -0,0 +1,204 @@
+import peergos.shared.util.*;
+
+import javax.crypto.*;
+import javax.crypto.spec.*;
+import java.security.*;
+import java.time.*;
+import java.util.*;
+import java.util.stream.*;
+
+public class S3Request {
+
+    /**
+     * Presign a url for a GET, PUT or POST
+     *
+     * @link https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+     */
+    public static PresignedUrl preSignUrl(String key,
+                                          int size,
+                                          String contentSha256,
+                                          boolean allowPublicReads,
+                                          ZonedDateTime now,
+                                          String verb,
+                                          String host,
+                                          Map<String, String> extraHeaders,
+                                          String region,
+                                          String accessKeyId,
+                                          String s3SecretKey) {
+        extraHeaders.put("Content-Length", "" + size);
+        S3Request policy = new S3Request(verb, host, key, contentSha256, allowPublicReads, extraHeaders,
+                accessKeyId, region, now.withNano(0).withZoneSameInstant(ZoneId.of("UTC")).toInstant());
+
+        String signature = computeSignature(policy, s3SecretKey);
+
+        return new PresignedUrl("https://" + host + "/" + key, policy.getHeaders(signature));
+    }
+
+    private static byte[] hmacSha256(byte[] secretKeyBytes, byte[] message) {
+        try {
+            Mac mac = Mac.getInstance("HMACSHA256");
+            SecretKey secretKey = new SecretKeySpec(secretKeyBytes, "HMACSHA256");
+            mac.init(secretKey);
+            return mac.doFinal(message);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static byte[] hmacSha256(String secretKey, byte[] message) {
+        return hmacSha256(secretKey.getBytes(), message);
+    }
+
+    /**
+     * Method for generating policy signature V4 for direct browser upload.
+     *
+     * @link https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html
+     */
+    public static String computeSignature(S3Request policy,
+                                          String s3SecretKey) {
+        String stringToSign = policy.stringToSign();
+        String shortDate = S3Request.asAwsShortDate(policy.date);
+
+        byte[] dateKey = hmacSha256("AWS4" + s3SecretKey, shortDate.getBytes());
+        byte[] dateRegionKey = hmacSha256(dateKey, policy.region.getBytes());
+        byte[] dateRegionServiceKey = hmacSha256(dateRegionKey, "s3".getBytes());
+        byte[] signingKey = hmacSha256(dateRegionServiceKey, "aws4_request".getBytes());
+
+        return ArrayOps.bytesToHex(hmacSha256(signingKey, stringToSign.getBytes()));
+    }
+
+    private static byte[] sha256(byte[] input) {
```

Comment:
Make public? As I can see the method has been copied elsewhere.

Reply:
I've removed it and used the existing method.